### DISCUSSION TITLE
Any plans to make Superset more DevOps-friendly?

### DISCUSSION THREAD

**Discussion by aaronsteers** [apache/superset#18208] [Q&A / Help]
Apologies if this is answered somewhere already but I'm new to Superset and ran into issues when trying to fit Superset into a DevOps framework with git and CI/CD.

A couple big blockers came up right away:

1. I couldn't find if the report definitions (and data source definitions) are stored anywhere in file form. There are some options to manually export them via the API, but the lack of access to the file directory makes any kind of commit back to git very cumbersome, and not realistic it seems for catching edits during report development in near real time.
2. The exported reports and definitions don't seem to be very friendly for a meaningful git diff. Ideally, these would be fairly readable so a pull request reviewer could understand the changes.
3. Also, following from 2, I'm worried if the same report is modified by two people in different branches, will I be able to merge those changes together after the fact, or will one set of changes end up clobbering the other?

Appreciate your thoughts on these issues in advance. I've heard great things about Superset and I know there's a huge community here. Has anyone found workaround for these issues, and/or are there development stories in the roadmap that might smooth out the DevOps lifecycle for reports and other Superset artifacts in the future?

Thanks!

**Comment by ad-m** [comment]
Do you know any good inspiration in that area?

I recognize Grafana filesystem provisioning ( https://grafana.com/docs/grafana/latest/administration/provisioning/ ) as an interesting replacement for API provisioning to make it more DevOps-friendly.

**Reply by siegenthalerroger** [reply]
A more cloud-native approach is the operator pattern, which when combined with gitops is very devops friendly. Grafana has a solution for this (finally) with the grafana-operator - https://github.com/grafana/grafana-operator. 

Essentially it just uses the API for provisioning with a well-defined lifecycle for resources.

**Comment by esselius** [comment]
I also had a very rough experience setting up superset in k8s. 

I use azure ad for SSO, [prometheus-statsd-exporter](https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus-statsd-exporter) to integrate with our existing monitoring stack.

Azure didn't work with openid auth for some reason and the clickhouse data source config requires me to perform a manual save through the UI (without changing anything) before it actually works, but I haven't had the time to figure out why 🤷 

As it took me quite a while to get data source definitions and SSO working, I'll share my `Dockerfile` and the relevant part of my `helmfile.yaml`

```
FROM apache/superset:b2bfb92cc4280d5bdc8dd98943db9746f777bbae

USER root
RUN --mount=type=cache,target=/root/.cache/pip pip install \
    psycopg2==2.9.1 \
    redis==3.5.3 \
    clickhouse-driver==0.2.2 \
    clickhouse-sqlalchemy==0.1.7 \
    elasticsearch-dbapi \
    pybigquery \
    shillelagh \
    prophet \
    lunarcalendar \
    Authlib \
    statsd
USER superset
```

```
  - name: superset
    namespace: team
    chart: superset/superset
    version: 0.5.6
    values:
      - image:
          repository: eu.gcr.io/employer/superset
          tag: {{ env "GITHUB_SHA" | default "latest" }}

        resources:
          requests:
            memory: 256Mi
            cpu: 100m

        ingress:
          enabled: true
          hosts:
            - superset.employer.dev
          tls:
            - secretName: superset-tls
              hosts:
                - superset.employer.dev
          annotations:
            kubernetes.io/ingress.class: cluster-wide
            cert-manager.io/cluster-issuer: letsencrypt-production

        envFromSecrets:
          - superset-azuread

        bootstrapScript: ""

        extraConfigs:
          import_datasources.yaml: |
            databases:
            - database_name: demo
              extra: '{"metadata_params":{},"engine_params":{},"metadata_cache_timeout":{},"schemas_allowed_for_csv_upload":[]}'
              sqlalchemy_uri: clickhouse+native://user:password@clickhouse-warehouse/businessmetrics
              tables: []

        configOverrides:
          featureFlags: |
            FEATURE_FLAGS = {
              "ALERT_REPORTS": True,
              "OMNIBAR": True,
              "DASHBOARD_NATIVE_FILTERS": True,
              "ROW_LEVEL_SECURITY": True,
              "DASHBOARD_RBAC": True,
            }
          auth: |
            from flask_appbuilder.security.manager import AUTH_OAUTH
            AUTH_TYPE = AUTH_OAUTH

            OAUTH_PROVIDERS = [
                {
                    "name": "azure",
                    "icon": "fa-sign-in",
                    "token_key": "access_token",
                    "remote_app": {
                        "client_id": os.environ.get("AZURE_APPLICATION_ID"),
                        "client_secret": os.environ.get("AZURE_SECRET"),
                        "api_base_url": "https://login.microsoftonline.com/<uuid>/oauth2",
                        "client_kwargs": {
                            "scope": "User.read name preferred_username email profile upn groups openid Groups.Read.All",
                            "resource": os.environ.get("AZURE_APPLICATION_ID"),
                        },
                        "request_token_url": None,
                        "access_token_url": "https://login.microsoftonline.com/<uuid>/oauth2/token",
                        "authorize_url": "https://login.microsoftonline.com/<uuid>/oauth2/authorize",
                    },
                },
            ]

            AUTH_USER_REGISTRATION = True
            AUTH_USER_REGISTRATION_ROLE = "Public"
            AUTH_ROLES_SYNC_AT_LOGIN = True

            ENABLE_PROXY_FIX = True

            from superset.security.manager import SupersetSecurityManager
            import logging
            logger = logging.getLogger("azure_ad")

            class AzureRoleBasedSecurityManager(SupersetSecurityManager):
                def _get_oauth_user_info(self, provider, resp):
                    if provider == "azure":
                        me = self._azure_jwt_token_parse(resp["id_token"])
                        logger.debug('User: "%s" has groups: %s', me["name"], ", ".join(me["groups"]))
                        return {
                            "name": me["name"],
                            "email": me["upn"],
                            "first_name": me.get("given_name", ""),
                            "last_name": me.get("family_name", ""),
                            "id": me["oid"],
                            "username": me["oid"],
                            "role_keys": me["groups"],
                        }
                    return {}
                oauth_user_info = _get_oauth_user_info

            CUSTOM_SECURITY_MANAGER = AzureRoleBasedSecurityManager

            AUTH_ROLES_MAPPING = {
              "<azure group uuid>": ["Data Team"],
            }

            FAB_ROLES = {
              "Data Team": [ ],
            }

            from superset.stats_logger import StatsdStatsLogger
            STATS_LOGGER = StatsdStatsLogger(host='prometheus-statsd-exporter', port=9125, prefix='superset')
```

**Comment by sycbbyes** [comment]
cannot load additional db driver correctly in K8S POD via formal docker image: apache/superset:2.0.1
because the K8S pod treate any application from docker image as stateless application.

steps：
1. install superset image on K8S
2. want to add elasticsearch db driver
3. access the linux os ssh , pip install elasticsearch-dbapi
4. but make no sense, becasue OS user superset cannot restart gunicorn application
5. while the K8S POD can only be replaced if restart it
6. at least, superset should provide a shell scripts to restart the superset application or allow user to pre-install necessary db driver

question:
is there any helpful way to restart the superset application via superset command?
or we can map one config file, so that the pip install list will cover any we need drivers easly?

**Reply by sycbbyes** [reply]
upon current K8S version 1.22,  application instance can still be accessable via docker command, while in the latest K8S version, it is seperated from docker, it will be more complex if there is any built in command support to restart superset.
